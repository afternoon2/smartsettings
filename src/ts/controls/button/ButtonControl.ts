import { Control } from '../control/Control';
import {
  ButtonControlProps, ControlListener,
} from '../control/Control.types';
import { InternalState } from '../../root/RootNode';

import Styles from '../../../sass/button.sass';

export class ButtonControl extends Control {
  public controlElement: HTMLButtonElement;

  protected static defaultType: string = 'button';

  protected static template = (state: InternalState): string => `
    <button
      class="${Styles.button}"
      type="${state.type || ButtonControl.defaultType}"
      title="${state.name}"
      ${state.disabled ? 'disabled="true"' : ''}
    >
      ${state.text || ''}
    </button>
  `;

  private controlListener?: ControlListener;
  private sectionListener?: ControlListener;
  private panelListener?: ControlListener;

  constructor(props: ButtonControlProps) {
    super(props, ButtonControl.template);
    this.controlElement = this.element.querySelector('button') as HTMLButtonElement;
    this.controlListener = props.listener;
    this.sectionListener = props.sectionListener;
    this.panelListener = props.panelListener;
    this.bindActionListeners();
  }

  private bindActionListeners() {
    this.controlElement.addEventListener('click', () => {
      if (this.controlListener) {
        this.controlListener({
          targetId: this.state.id,
          key: '',
          value: '',
          listenerType: 'control'
        });
      }
      if (this.sectionListener) {
        this.sectionListener({
          targetId: this.state.id,
          key: '',
          value: '',
          listenerType: 'section',
        });
      }
      if (this.panelListener) {
        this.panelListener({
          targetId: this.state.id,
          key: '',
          value: '',
          listenerType: 'panel'
        });
      }
    });
  }
}
