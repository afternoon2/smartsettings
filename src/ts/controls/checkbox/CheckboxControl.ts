import cuid from 'cuid';

import { Control } from '../control/Control';
import { InternalState, Listener, ListenerUpdate, CheckboxControlOptions } from '../../types';

import Styles from '../../../sass/checkbox.sass';

export type CheckboxControlProps = {
  id: string,
  options: CheckboxControlOptions,
  parentElement: HTMLElement,
  sectionListener?: Listener,
  panelListener?: Listener,
};

export class CheckboxControl extends Control {
  public controlElement: HTMLInputElement;
  public readonly displayType = 'checkbox';

  protected static template = (state: InternalState): string => {
    const id: string = cuid();
    return `<label
      class="${Styles.checkbox}" for="${id}">
        <input
          id="${id}"
          type="checkbox"
          title="${state.name}"
          ${state.checked ? 'checked' : ''}
          ${state.disabled ? 'disabled' : ''}
          ${state.readonly ? 'readonly="true"' : ''}
        />
      </label>
    `;
  }

  constructor(props: CheckboxControlProps) {
    super({
      ...props,
      template: CheckboxControl.template,
    });
    this.controlElement = this.element.querySelector('[type="checkbox"]') as HTMLInputElement;
    this.listeners.set('checked', this.onCheck);
    this.bindActionListeners();
  }

  get checked(): boolean {
    return this.state.checked as boolean;
  }

  check() {
    this.state.checked = true;
  }

  uncheck() {
    this.state.checked = false;
  }

  toggle() {
    this.state.checked = !this.state.checked;
  }

  private onCheck: Listener = (update: ListenerUpdate) => {
    if (update.value === true) {
      this.controlElement.setAttribute('checked', 'true');
    } else {
      this.controlElement.removeAttribute('checked');
    }
  }

  private bindActionListeners() {
    this.controlElement.addEventListener('change', (event: Event) => {
      event.preventDefault();
      this.toggle();
    });
  }
}
