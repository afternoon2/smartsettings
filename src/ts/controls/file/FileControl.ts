import cuid from 'cuid';

import { Control } from '../control/Control';
import { InternalState, Listener, ListenerUpdate, FileControlOptions } from '../../types';

import Styles from '../../../sass/file.sass';

export type FileControlProps = {
  id: string,
  options: FileControlOptions,
  parentElement: HTMLElement,
  sectionListener?: Listener,
  panelListener?: Listener,
};

export class FileControl extends Control {
  public controlElement: HTMLInputElement;

  protected static template = (state: InternalState): string => {
    const id: string = cuid();
    return `<label
        class="${Styles.file}"
        for="${id}"
      >
        Upload...
        <input
          id="${id}"
          type="file"
          title="${state.name}"
          ${state.disabled ? 'disabled' : ''}
          ${state.readonly ? 'readonly="true"' : ''}
        />
      </label>
    `
  };

  constructor(props: FileControlProps) {
    super({
      ...props,
      template: FileControl.template,
    });
    this.controlElement = this.element.querySelector('[type="file"]') as HTMLInputElement;
    this.listeners.set('accept', this.onAcceptChange);
    this.bindActionListeners();
  }

  get accept(): string {
    return this.state.accept as string;
  }

  set accept(accept: string) {
    this.state.accept = accept;
  }

  get files(): FileList | null {
    return this.state.files as FileList;
  }

  private onAcceptChange: Listener = (update: ListenerUpdate) => {
    this.controlElement.setAttribute('accept', update.value as string);
  }

  private bindActionListeners() {
    this.controlElement.addEventListener('change', (event: Event) => {
      this.state.files = ((event as Event).target as HTMLInputElement).files as FileList;
    });
  }
}
