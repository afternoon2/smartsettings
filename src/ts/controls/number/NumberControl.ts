import { Control } from '../control/Control';
import {
  NumberControlProps, ControlListener, ControlListenerUpdate,
} from '../control/Control.types';
import { InternalState } from '../../root/RootNode';

import Styles from '../../../sass/number.sass';

export class NumberControl extends Control {
  public controlElement: HTMLInputElement;

  protected static template = (state: InternalState): string => `<input
      type="number"
      class="${Styles.number}"
      ${state.value ? `value="${state.value}"` : ''}
      ${typeof state.min === 'number' ? `min="${state.min}"` : ''}
      ${typeof state.max === 'number' ? `max="${state.max}"` : ''}
      ${typeof state.step === 'number' ? `step="${state.step}"` : ''}
      ${state.readonly ? 'readonly="true"' : ''}
      ${state.disabled ? 'disabled' : ''}
      ${state.placeholder ? `placeholder="${state.placeholder}"` : ''}
    />
  `;

  constructor(props: NumberControlProps) {
    super(props, NumberControl.template);
    this.controlElement = this.element.querySelector('[type="number"]') as HTMLInputElement;
    this.listeners.set('value', this.onValue);
    this.bindActionListeners();
  }

  set value(value: number) {
    this.state.value = value;
  }

  get value(): number {
    return this.state.value as number;
  }

  private onValue: ControlListener = (update: ControlListenerUpdate) => {
    this.controlElement.value = update.value as string;
  }

  private bindActionListeners = () => {
    this.controlElement.addEventListener('change', (e: Event) => {
      this.value = parseFloat((e.target as HTMLInputElement).value);
    });
  }
}