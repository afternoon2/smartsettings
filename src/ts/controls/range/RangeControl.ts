import { Control } from '../control/Control';
import {
  RangeControlProps, ControlListener, ControlListenerUpdate,
} from '../control/Control.types';
import { InternalState } from '../../root/RootNode';

import Styles from '../../../sass/range.sass';
import ControlStyles from '../../../sass/control.sass';

export class RangeControl extends Control {
  public controlElement: HTMLInputElement;

  protected static template = (state: InternalState): string => `<input
      type="range"
      class="${Styles.range}"
      ${state.value && typeof state.value === 'number' ? `value="${state.value}"` : ''}
      ${typeof state.min === 'number' ? `min="${state.min}"` : ''}
      ${typeof state.max === 'number' ? `max="${state.max}"` : ''}
      ${typeof state.step === 'number' ? `step="${state.step}"` : ''}
      ${state.readonly ? 'readonly' : ''}
      ${state.disabled ? 'disabled' : ''}
    />
  `;

  constructor(props: RangeControlProps) {
    super(props, RangeControl.template);
    this.controlElement = this.element.querySelector('[type="range"]') as HTMLInputElement;
    this.updateMain();
    this.listeners.set('value', this.onValue);
    this.bindActionListeners();
  }

  set value(value: number) {
    this.state.value = value;
  }

  get value(): number {
    return this.state.value as number;
  }

  private onValue: ControlListener = (update: ControlListenerUpdate) => {
    this.controlElement.value = update.value as string;
    (this.element.querySelector(`.${Styles.rangeValue}`) as HTMLSpanElement).textContent = update.value as string;
  }

  private bindActionListeners = () => {
    this.controlElement.addEventListener('input', (e: Event) => {
      this.value = parseFloat((e.target as HTMLInputElement).value);
    });
  }

  private updateMain = () => {
    const main = this.element.querySelector(`.${ControlStyles.control__content}`) as HTMLElement;
    main.classList.add(`${ControlStyles['control__content--range']}`);
    main.insertAdjacentHTML('beforeend', `<span class="${Styles.rangeValue}">
      ${typeof this.state.value === 'number' ? this.state.value : ''}
    </span>`);
  }
}