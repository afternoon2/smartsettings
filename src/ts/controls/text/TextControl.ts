import { Control } from '../control/Control';
import { InternalState, TextControlOptions, Listener, ListenerUpdate } from '../../types';

import Styles from '../../../sass/text.sass';

export type TextControlProps = {
  id: string,
  options: TextControlOptions,
  parentElement: HTMLElement,
  sectionListener?: Listener,
  panelListener?: Listener,
};

export class TextControl extends Control {
  public controlElement: HTMLInputElement;
  public readonly displayType = 'text';

  protected static template = (state: InternalState): string => `
    <input
      type="text"
      class="${Styles.text}"
      ${state.disabled ? 'disabled' : ''}
      ${state.value ? `value="${state.value}"` : ''}
      ${state.autocomplete ? 'autocomplete="true"' : ''}
      ${state.minLength ? `min-length="${state.minLength}"` : ''}
      ${state.maxLength ? `min-length="${state.maxLength}"` : ''}
      ${state.placeholder ? `placeholder="${state.placeholder}"` : ''}
      ${state.pattern ? `pattern="${state.pattern}"` : ''}
      ${state.readOnly ? 'readonly="true"' : ''}
    />
  `;

  constructor(props: TextControlProps) {
    super({
      ...props,
      template: TextControl.template
    });
    this.controlElement = this.element.querySelector('[type="text"]') as HTMLInputElement;
    this.listeners.set('value', this.onValue);
    this.bindActionListeners();
  }

  set value(value: string) {
    this.state.value = value;
  }

  get value(): string {
    return this.state.value as string;
  }

  private onValue: Listener = (update: ListenerUpdate) => {
    this.controlElement.value = update.value as string;
  }

  private bindActionListeners = () => {
    this.controlElement.addEventListener('input', (e: Event) => {
      this.value = (e.target as HTMLInputElement).value;
    });
  }
}
