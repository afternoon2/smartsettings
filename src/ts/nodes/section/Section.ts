import { SectionProps } from '../nodes.types';
import { ParentNode } from '../parent/ParentNode';
import { InternalState } from '../../root/RootNode';

import Base from '../../../sass/base.sass';
import Styles from '../../../sass/section.sass';

export class SectionNode extends ParentNode {
  public headerElement: HTMLElement;
  public bodyElement: HTMLElement;

  protected static template = (state: InternalState): string => `<header
      class="${Styles.section__header}"
    >
      <p class="${Styles.section__name}">
        ${state.name}
      </p>
      <a class="${Styles.section__link}">
        ${state.collapsed ? '&#9662;' : '&#9656;'}
      </a>
    </header>
    <main class="${Styles.section__body} ${state.collapsed ? Base.hidden : ''}"></main>
  `;

  constructor(props: SectionProps) {
    super(props, SectionNode.template);
    this.fillInElement(Styles.section);
    this.headerElement = this.element.querySelector(`.${Styles.section__header}`) as HTMLElement;
    this.bodyElement = this.element.querySelector(`.${Styles.section__body}`) as HTMLElement;
    this.bindEventListeners();
  }

  private onHeaderClick = (event: Event) => {
    event.preventDefault();
    if (this.state.collapsed === true) {
      this.open();
    } else {
      this.close();
    }
  }

  private bindEventListeners() {
    this.headerElement.addEventListener('click', this.onHeaderClick);
  }
}