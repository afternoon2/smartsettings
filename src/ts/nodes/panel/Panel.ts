import cuid from 'cuid';

import { ParentNode, AnyControl } from '../parent/ParentNode';
import { PanelProps, PanelPosition, SectionOptions } from '../nodes.types';
import { InternalState } from '../../root/RootNode';

import Base from '../../../sass/base.sass';
import Styles from '../../../sass/panel.sass';
import { ControlListener, ControlListenerUpdate, ControlOptions } from '../../controls/control/Control.types';
import { SectionNode } from '../section/Section';

export class PanelNode extends ParentNode {
  public bodyElement: HTMLElement;
  public headerElement: HTMLElement;
  private toggleElement: HTMLAnchorElement;
  private nameElement: HTMLParagraphElement;

  protected static template = (state: InternalState): string => `<header
      class="${Styles.panel__header}"
      ${state.draggable ? 'draggable="true"' : ''}
    >
      <p class="${Styles.panel__name}" title="${state.name}">
        ${state.name}
      </p>
      <a class="${Styles.panel__link}">
        ${state.collapsed ? '&#9662;' : '&#9656;'}
      </a>
    </header>
    <main class="${Styles.panel__body} ${state.collapsed ? Base.hidden : ''}">
    </main>`;

  constructor(props: PanelProps) {
    super(props, PanelNode.template);
    this.fillInElement(Styles.panel);
    this.headerElement = this.element.querySelector(`.${Styles.panel__header}`) as HTMLElement;
    this.nameElement = this.headerElement.querySelector('p') as HTMLParagraphElement;
    this.toggleElement = this.headerElement.querySelector('a') as HTMLAnchorElement;
    this.bodyElement = this.element.querySelector(`.${Styles.panel__body}`) as HTMLElement;
    this.listeners.set('top', this.onPositionChanged);
    this.listeners.set('left', this.onPositionChanged);
    if (props.listener) {
      this.listeners.set('panel', props.listener);
    }
    this.bindEventListeners();
    if (!this.state.top || !this.state.left) {
      this.setPosition({
        top: 10,
        left: 10,
      });
    }
  }

  get position(): PanelPosition | null {
    return this.state.top && this.state.left ?
      {
        top: this.state.top as number,
        left: this.state.left as number,
      } :
      null;
  }

  setPosition(position: PanelPosition) {
    this.state.top = position.top;
    this.state.left = position.left;
  }

  setListener(listener: ControlListener) {
    this.listeners.set('panel', listener);
  }

  destroy() {
    this.registry.clear();
    this.listeners.clear();
    this.parentElement.removeChild(this.element);
  }

  remove(name: string) {
    const toRemove = Array.from(this.registry.values())
      .find((el: AnyControl | SectionNode) => el.name === name);
    if (!toRemove) {
      this.registry.forEach((value: AnyControl | SectionNode) => {
        if (value instanceof SectionNode) {
          value.remove(name);
        };
      });
    } else {
      this.bodyElement.removeChild(toRemove.element);
      this.registry.delete(toRemove.id);
    }
  }

  removeById(id: string) {
    const localElement = this.registry.get(id);
    if (localElement) {
      this.bodyElement.removeChild(localElement.element);
      this.registry.delete(id);
    } else {
      this.registry.forEach((value: AnyControl | SectionNode) => {
        if (value instanceof SectionNode) {
          value.removeById(id);
        }
      })
    }
  }

  removeAll() {
    while (this.bodyElement.firstChild) {
      this.bodyElement.removeChild(
        this.bodyElement.firstChild,
      );
    }
    this.registry.clear();
  }

  control(
    control: string,
    name: string,
    options: ControlOptions | null,
    listener?: ControlListener,
  ): AnyControl | null {
    return this.createControl(
      control,
      name,
      options,
      {
        control: listener,
        panel: this.listeners.get('panel'),
      },
    );
  }

  section(
    name: string, 
    options: SectionOptions, 
    listener?: ControlListener
  ): SectionNode {
    const id: string = cuid();
    const section = new SectionNode({
      id,
      name,
      options,
      parentElement: this.bodyElement,
      listener,
      panelListener: this.listeners.get('panel'),
    });
    this.registry.set(id, section);
    return section;
  }

  private bindEventListeners() {
    this.toggleElement.addEventListener('click', this.toggleClickListener);
    this.nameElement.addEventListener('click', this.toggleClickListener);
  }

  private toggleClickListener = (event: Event) => {
    event.preventDefault();
    if (this.state.collapsed === true) {
      this.open();
    } else {
      this.close();
    }
  }

  private onPositionChanged: ControlListener = (update: ControlListenerUpdate) => {
    const cssProp: string = update.key;
    // @ts-ignore
    this.element.style[cssProp] = `${update.value}px`;
  }
}
