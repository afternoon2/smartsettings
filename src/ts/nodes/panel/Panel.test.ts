import { PanelNode } from './Panel';
import { ButtonControl } from '../../controls/button/ButtonControl';
import { TextControl } from '../../controls/text/TextControl';
import { TextAreaControl } from '../../controls/textarea/TextAreaControl';
import { NumberControl } from '../../controls/number/NumberControl';
import { RangeControl } from '../../controls/range/RangeControl';
import { CheckboxControl } from '../../controls/checkbox/CheckboxControl';
import { FileControl } from '../../controls/file/FileControl';
import { SectionNode } from '../section/Section';
import { PanelPosition } from '../../types';

let panel: PanelNode;
let listener: any;

describe('Panel node', () => {
  beforeEach(() => {
    listener = jest.fn();
    panel = new PanelNode({
      name: 'name',
      collapsed: false,
      disabled: false,
      invisible: false,
      listener,
    });
  });

  afterEach(() => {
    document.body.innerHTML = '';
  });

  test('Open and close', () => {
    panel.close();
    expect(panel.collapsed).toBe(true);
    panel.open();
    expect(panel.collapsed).toBe(false);
    expect(listener).toHaveBeenCalledTimes(2);
  });

  test('Open and close by click', () => {
    (panel.headerElement.querySelector('a') as HTMLAnchorElement).click();
    expect(panel.collapsed).toBe(true);
  });

  test('Control creation', () => {
    const btn = panel.control('button', {
      name: 'Button',
    });
    const text = panel.control('text', {
      name: 'Text',
      placeholder: 'Write here!',
      value: 'text input',
    });
    const textarea = panel.control('textarea', {
      name: 'Textarea',
      placeholder: 'textarea',
      value: 'Textarea',
    });
    const range = panel.control('range', {
      name: 'Range',
      min: 0, max: 10, value: 1, step: 1,
    });
    const number = panel.control('number', {
      name: 'Number',
      min: 0, max: 10, value: 1, step: 1,
    });
    const checkbox = panel.control('checkbox', {
      name: 'Checkbox',
      checked: true,
    });
    const file = panel.control('file', {
      name: 'File',
    });
    const falsyControl = panel.control('falsy', {
      name: 'Falsy',
      min: 1, max: 10, step: 1, value: 10,
    });
    expect(btn).toBeInstanceOf(ButtonControl);
    expect(text).toBeInstanceOf(TextControl);
    expect(textarea).toBeInstanceOf(TextAreaControl);
    expect(number).toBeInstanceOf(NumberControl);
    expect(range).toBeInstanceOf(RangeControl);
    expect(checkbox).toBeInstanceOf(CheckboxControl);
    expect(file).toBeInstanceOf(FileControl);
    expect(falsyControl).toBe(null);
  });

  test('Section creation', () => {
    const section = panel.section({
      name: 'Section'
    });
    expect(section).toBeInstanceOf(SectionNode);
    expect(section.name).toBe('Section');
  });

  test('Set and get panel position', () => {
    panel.setPosition({
      top: 30,
      left: 40,
    });
    expect(panel.element.style.top).toBe('30px');
    expect(panel.element.style.left).toBe('40px');
    expect((panel.position as PanelPosition).left).toBe(40);
    expect((panel.position as PanelPosition).top).toBe(30);
  });

  test('Destroy panel', () => {
    panel.destroy();
    expect(document.getElementById(panel.id)).toBe(null);
    expect(panel.parentElement.querySelector(`#${panel.id}`)).toBe(null);
    
    expect(() => {
      panel.rename('Test');
    }).toThrowError();
  });

  test('Remove method', () => {
    panel.control('button', { name: 'Button' });
    panel.control('button2', { name: 'Button' });
    panel.remove('Button');
    expect(panel.bodyElement.querySelector('#button')).toBe(null);
    expect(panel.bodyElement.querySelector('#button2')).toBe(null);

    const section = panel.section({
      name: 'Section',
    });
    const nestedButton = section.control('button', {
      name: 'Button in section',
    }) as ButtonControl;

    panel.remove('Button in section');

    expect(section.bodyElement.querySelector(nestedButton.id)).toBe(null);
  });

  test('Remove by id method', () => {
    const section = panel.section({ name: 'Section' });
    const button = panel.control('button', { name: 'Button' }) as ButtonControl;
    const nestedButton = section.control('button', { name: 'Button in section' }) as ButtonControl;

    panel.removeById(button.id);
    expect(panel.bodyElement.querySelector(`#${button.id}`)).toBe(null);

    panel.removeById(nestedButton.id);
    expect(section.bodyElement.querySelector(`#${nestedButton.id}`)).toBe(null);

  });

  test('Remove all', () => {
    const section = panel.section({ name: 'Section' });
    panel.control('button', { name: 'Button' }) as ButtonControl;
    section.control('button', { name: 'Button in section' }) as ButtonControl;
    panel.removeAll();

    expect(panel.bodyElement.childElementCount).toBe(0);
  });
});

