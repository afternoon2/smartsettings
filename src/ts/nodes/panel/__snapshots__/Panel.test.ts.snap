// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel node If it matches the snapshot 1`] = `
<div
  class="panel"
  style="top: 10px; left: 10px;"
>
  <header
    class="panel__header"
  >
    
      
    <p
      class="panel__name"
      title="name"
    >
      
        name
      
    </p>
    
      
    <a
      class="panel__link"
    >
      
        ▸
      
    </a>
    
    
  </header>
  
    
  <main
    class="panel__body "
  >
    
    
  </main>
</div>
`;

exports[`Panel node Open and close 1`] = `
<div
  class="panel"
  style="top: 10px; left: 10px;"
>
  <header
    class="panel__header"
  >
    
      
    <p
      class="panel__name"
      title="name"
    >
      
        name
      
    </p>
    
      
    <a
      class="panel__link"
    >
      ▾
    </a>
    
    
  </header>
  
    
  <main
    class="panel__body hidden"
  >
    
    
  </main>
</div>
`;

exports[`Panel node Open and close 2`] = `
<div
  class="panel"
  style="top: 10px; left: 10px;"
>
  <header
    class="panel__header"
  >
    
      
    <p
      class="panel__name"
      title="name"
    >
      
        name
      
    </p>
    
      
    <a
      class="panel__link"
    >
      ▸
    </a>
    
    
  </header>
  
    
  <main
    class="panel__body"
  >
    
    
  </main>
</div>
`;

exports[`Panel node Open and close by click 1`] = `
<div
  class="panel"
  style="top: 10px; left: 10px;"
>
  <header
    class="panel__header"
  >
    
      
    <p
      class="panel__name"
      title="name"
    >
      
        name
      
    </p>
    
      
    <a
      class="panel__link"
    >
      ▾
    </a>
    
    
  </header>
  
    
  <main
    class="panel__body hidden"
  >
    
    
  </main>
</div>
`;
